### PR TITLE
[HLSL] Implementation of the frac intrinsic

### PR DESCRIPTION
**Pull request by farzonl**
This change implements the frontend  for #70099
Builtins.td           - add the frac builtin
CGBuiltin.cpp         - add the builtin to DirectX intrinsic mapping
hlsl_intrinsics.h     - add the frac api
SemaChecking.cpp      - add type checks for builtin
IntrinsicsDirectX.td  - add the frac intrinsic

The backend changes for this are going to be very simple: https://github.com/llvm/llvm-project/commit/f309a0eb558b65dfaff0d1d23b7d07fb07e27121

They were not included because llvm/lib/Target/DirectX/DXIL.td is going through a major refactor.

### CODE REVIEW

File: clang/test/SemaHLSL/BuiltIns/frac-errors.hlsl [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm -disable-llvm-passes -verify -verify-ignore-unexpected
+
```
Comment (python3kgae) [inline clang/test/SemaHLSL/BuiltIns/frac-errors.hlsl:3]: What will happen for 
```
struct A { float a; }; 
A fa;
__builtin_hlsl_elementwise_frac (fa)
```

Comment (farzonl) [reply clang/test/SemaHLSL/BuiltIns/frac-errors.hlsl:3]: error: 1st argument must be a vector, integer or floating point type (was 'A')

Comment (python3kgae) [reply clang/test/SemaHLSL/BuiltIns/frac-errors.hlsl:3]: Just curious.
Don't need a test case I think.

**Review by python3kgae — APPROVED** [review]

File: clang/test/CodeGenHLSL/builtins/frac.hlsl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+// NATIVE_HALF: define noundef half @
+// NATIVE_HALF: %dx.frac = call half @llvm.dx.frac.f16(
+// NATIVE_HALF: ret half %dx.frac
+// NO_HALF: define noundef float @"?test_frac_half@@YA$halff@$halff@@Z"(
+// NO_HALF: %dx.frac = call float @llvm.dx.frac.f32(
+// NO_HALF: ret float %dx.frac
+half test_frac_half ( half p0 ) {
+  return frac ( p0 );
+}
```
Comment (bogner) [inline clang/test/CodeGenHLSL/builtins/frac.hlsl:17]: Please follow LLVM formatting conventions in test cases, ie:
```c++
half test_frac_half(half p0) {
  return frac (p0);
}
```
(aside: we should probably teach clang-format to recognize hlsl)

**Review by bogner — APPROVED** [review]
LGTM other than the formatting in the test cases

File: clang/lib/Sema/SemaChecking.cpp [hunk 3]
```diff
@@ -5246,6 +5246,23 @@ bool CheckVectorElementCallArgs(Sema *S, CallExpr *TheCall) {
   return true;
 }
 
+bool CheckAllArgsAreFloatRepresentation(Sema *S, CallExpr *TheCall) {
```
Comment (bharadwajy) [inline clang/lib/Sema/SemaChecking.cpp:5249]: Consider changing the function name to either
`CheckAllArgsAreFloatType`, `CheckAllArgsAreFloats` or `CheckAllArgsHaveFloatRepresentation`

**Pull request merged**